providers/common/kdfs/hkdf.c

```diff
@@ -35,209 +46,244 @@ static int HKDF_Expand(const EVP_MD *evp_md,
                        const unsigned char *info, size_t info_len,
                        unsigned char *okm, size_t okm_len);
 
-struct evp_kdf_impl_st {
+typedef struct {
+    void *provctx;
     int mode;
-    const EVP_MD *md;
+    EVP_MD *md;
     unsigned char *salt;
     size_t salt_len;
     unsigned char *key;
     size_t key_len;
     unsigned char info[HKDF_MAXBUF];
     size_t info_len;
-};
+} KDF_HKDF;
 
-static EVP_KDF_IMPL *kdf_hkdf_new(void)
+static void *kdf_hkdf_new(void *provctx)
 {
-    EVP_KDF_IMPL *impl;
+    KDF_HKDF *ctx;
 
-    if ((impl = OPENSSL_zalloc(sizeof(*impl))) == NULL)
-        KDFerr(KDF_F_KDF_HKDF_NEW, ERR_R_MALLOC_FAILURE);
-    return impl;
+    if ((ctx = OPENSSL_zalloc(sizeof(*ctx))) == NULL)
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+    else
+        ctx->provctx = provctx;
+    return ctx;
 }
 
-static void kdf_hkdf_free(EVP_KDF_IMPL *impl)
+static void kdf_hkdf_free(void *vctx)
 {
-    kdf_hkdf_reset(impl);
-    OPENSSL_free(impl);
-}
+    KDF_HKDF *ctx = (KDF_HKDF *)vctx;
 
-static void kdf_hkdf_reset(EVP_KDF_IMPL *impl)
-{
-    OPENSSL_free(impl->salt);
-    OPENSSL_clear_free(impl->key, impl->key_len);
-    OPENSSL_cleanse(impl->info, impl->info_len);
-    memset(impl, 0, sizeof(*impl));
+    kdf_hkdf_reset(ctx);
+    EVP_MD_meth_free(ctx->md);
```

Comment:
If you do `EVP_MD_meth_free()` before `kdf_hkdf_reset()`, you won't leak an MD...